### ISSUE TITLE
Deny connections from multiple devices for the same account and OTP code (unique PIN code)

### ISSUE THREAD

**Issue by githubcom13** [PHPAuth/PHPAuth#413]
Hi,
- is it possible to add a function for deny multiple connections from multiple devices?
- Is it possible to add an OTP (One-Time PIN) Code sending function ?
Regards

**Comment by louis123562** [comment]
> Is it possible to add an OTP (One-Time PIN) Code sending mail function ?
Sure - fork yourself the repo and implement it.

> Is it possible to add a user notification function by email for each successful or failed connection?
What do you mean exactly?

**Comment by githubcom13** [comment]
I'm not using your class, it's just suggestions for improvement.
Your class is focused on security, but it does not use simple protection mechanisms such as OTP code or suspicious activity notification emails. Having chopped passwords does not help without a good security mechanism.  The big actors of the Web like Facebook, Twitter, Github, Digitalocean, etc., use its security mechanisms. Bonne continuation, merci pour le partage. Regards

**Comment by louis123562** [comment]
OTP is without a service prodiver is much more work. We could send suspicious e-mail notifications, but it's on your own.

**Comment by Conver** [comment]
I did add OTP but I did also remove it again, see past commits.

**Comment by githubcom13** [comment]
OTP can be sent by email as on Github.
Best regards

**Comment by Conver** [comment]
e-mail notifications would be very easy to add tho.

**Comment by louis123562** [comment]
Sorry, i misunderstood. OTP as a 2FA is without a second server not very safe?!

**Comment by githubcom13** [comment]
I have the impression that you did not understand the OTP code mechanism. An OTP code can be sent by email / SMS `user -> auth -> sendOTPCode -> checkOTPCode -> login`. You can generate an OTP code like this ```
// Generate 6 digit OTP code
    public function createOTP(): int {
        return random_int(100000, 999999);
    }
```. It is a PIN code. You can also use question / answer for the password reset request. This class is not secure insofar as it uses a low level security system (login basic + hash password)...

**Comment by louis123562** [comment]
I dont get it. For passwort reset, we already use a key, which is much more complex than the OTP you mentioned.

**Comment by githubcom13** [comment]
No, OTP code should be used at each login, if the attacker is in poced of the password of the user he cannot connect without indicating the Pin code (OTP code) received by mail / sms.
The question / answer is to be used only for the request to reset the password. His questions / answers you must collect at the time of registration (eg what was the name of your first pet ...) alone, the user must know the answers, if the user's email address was compromised so the attacker must know the answers to reset the password. It would also be good to store the session_id in the database, at each connection you make a comparison and disconnect the user with the old session, this avoids sharing login/password. Sorry for my English. best regards

**Comment by louis123562** [comment]
okay, makes sense. But i would not always send the OTP, only when the user's IP or agent changes he would need an OTP.

**Comment by githubcom13** [comment]
Yes, it is up to you to define the level of your security policy. Best regards

**Comment by Conver** [comment]
I'm adding OTP/2FA in one of my project, I'll merge the code into PHPAuth soon.

**Comment by githubcom13** [comment]
Good job, thanks for follow.

**Comment by Conver** [comment]
I was to fast, putting this on hold until we can figure out a way to make it as a plugin.

**Comment by Conver** [comment]
Update: I will add it into the base code as I think it's important security feature to have, but I will make it configurable with a few settings to fit different levels of needs.

**Comment by Conver** [comment]
@all-contributors please add Conver for Code

**Comment by Criptos** [comment]
I have just implemented OTP for one of my projects with: 
https://github.com/Spomky-Labs/otphp

**Comment by jmhumphrey** [comment]
This is what I did.  I used [fingerprint2js](https://fingerprintjs.com) on the front end to generate a unique device hash.  Using the http_user_agent also would work but i prefer the hash personally.  See attached for my 2fa / device registration class extension.

[CustomAuth.php.txt](https://github.com/PHPAuth/PHPAuth/files/5397961/CustomAuth.php.txt)
